src/go/k8s/controllers/redpanda/console_controller.go

```diff
@@ -0,0 +1,219 @@
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+package redpanda
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	redpandav1alpha1 "github.com/redpanda-data/redpanda/src/go/k8s/apis/redpanda/v1alpha1"
+	adminutils "github.com/redpanda-data/redpanda/src/go/k8s/pkg/admin"
+	consolepkg "github.com/redpanda-data/redpanda/src/go/k8s/pkg/console"
+	"github.com/redpanda-data/redpanda/src/go/k8s/pkg/resources"
+)
+
+// ConsoleReconciler reconciles a Console object
+type ConsoleReconciler struct {
+	client.Client
+	Scheme                *runtime.Scheme
+	Log                   logr.Logger
+	AdminAPIClientFactory adminutils.AdminAPIClientFactory
+	clusterDomain         string
+	Store                 *consolepkg.Store
+	EventRecorder         record.EventRecorder
+}
+
+const (
+	// Warning event if referenced Cluster not found
+	ClusterNotFoundEvent = "ClusterNotFound"
+
+	// Warning event if subdomain is not found in Cluster ExternalListener
+	NoSubdomainEvent = "NoSubdomain"
+)
+
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apps,resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;update;patch
+
+//+kubebuilder:rbac:groups=redpanda.vectorized.io,resources=consoles,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=redpanda.vectorized.io,resources=consoles/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=redpanda.vectorized.io,resources=consoles/finalizers,verbs=update
+
+func (r *ConsoleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("redpandaconsole", req.NamespacedName)
+
+	log.Info(fmt.Sprintf("Starting reconcile loop for %v", req.NamespacedName))
+	defer log.Info(fmt.Sprintf("Finished reconcile loop for %v", req.NamespacedName))
+
+	console := &redpandav1alpha1.Console{}
+	if err := r.Get(ctx, req.NamespacedName, console); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	cluster := &redpandav1alpha1.Cluster{}
+	if err := r.Get(ctx, console.GetClusterRef(), cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Console will never reconcile if Cluster is not found
+			// Users shouldn't check logs of operator to know this
+			// Adding Conditions in Console status might not be apt, record Event instead
+			// Alternatively, we can have this validation via Webhook
+			r.EventRecorder.Eventf(
```

Comment:
is this going to forever write new events like this to the event log?

Reply:
It will use the default retry backoff. K8s events are not persisted, only recent events are shown.

Reply:
this is actually not true, events are in etcd https://github.com/kubernetes/kubernetes/issues/4432#issuecomment-549953993

Reply:
I apologize, my wording is wrong :sweat_smile: They are indeed persisted and old events are garbage collected. Only the recent ones (by default 1 hour) are in etcd I think. As discussed, we'll leave it as is instead of logging the error in the operator as this seem a good use case of using events to inform the user of relevant info. :pray: